### PR TITLE
Add CommandBlocks

### PR DESCRIPTION
**Pull request by Snaipe**
I had a little issue with git and I have to send a second PR, sorry..
Signed-off-by: Snaipe franklinmathieu@gmail.com

### CODE REVIEW

**Comment by Snaipe** [comment]
The PR comes with an other PR for Spout : https://github.com/SpoutDev/Spout/pull/2165

File: src/main/java/org/spout/vanilla/plugin/protocol/msg/world/block/CommandBlockMessage.java [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+	
+	public int getZ() {
+		return z;
+	}
+	
+	public String getCommandInput() {
+		return command;
+	}
+	
+	public Message wrap(boolean upstream, Protocol activeProtocol) {
```
Comment (zml2008) [inline src/main/java/org/spout/vanilla/plugin/protocol/msg/world/block/CommandBlockMessage.java]: This is unnecessary

File: src/main/java/org/spout/vanilla/plugin/protocol/msg/world/block/CommandBlockMessage.java [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+ * more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License,
+ * the MIT license and the Spout License Version 1 along with this program.
+ * If not, see <http://www.gnu.org/licenses/> for the GNU Lesser General Public
+ * License and see <http://spout.in/licensev1> for the full license, including
+ * the MIT license.
+ */
+/*
+ * This file is part of SpoutAPI.
```
Comment (zml2008) [inline src/main/java/org/spout/vanilla/plugin/protocol/msg/world/block/CommandBlockMessage.java]: Double license block

File: src/main/java/org/spout/vanilla/plugin/protocol/handler/world/block/CommandBlockHandler.java [hunk 3]
```diff
@@ -24,12 +24,41 @@
...
+		
+		final Player player = session.getPlayer();
+		
+		int x = message.getX();
+		int y = message.getY();
+		int z = message.getZ();
+		String command = message.getCommandInput();
+		
+		String[] split = command.split(" ");
+		if (split.length == 0 || !player.hasPermission("vanilla.commandblock." + split[0])) {
```
Comment (zml2008) [inline src/main/java/org/spout/vanilla/plugin/protocol/handler/world/block/CommandBlockHandler.java]: Trim the / of of this (if present)

File: src/main/java/org/spout/vanilla/plugin/protocol/codec/world/block/CommandBlockCodec.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+import org.jboss.netty.buffer.ChannelBuffer;
+import org.jboss.netty.buffer.ChannelBuffers;
+
+import org.spout.api.protocol.MessageCodec;
+import org.spout.api.util.Named;
+import org.spout.vanilla.plugin.protocol.ChannelBufferUtils;
+import org.spout.vanilla.plugin.protocol.msg.world.block.CommandBlockMessage;
+
+public class CommandBlockCodec extends MessageCodec<CommandBlockMessage> implements Named {
+	public static final String channelName = "MC|AdvCdm";
```
Comment (zml2008) [inline src/main/java/org/spout/vanilla/plugin/protocol/codec/world/block/CommandBlockCodec.java]: this should be CHANNEL_NAME

File: src/main/java/org/spout/vanilla/plugin/protocol/VanillaHandlerLookupService.java [hunk 5]
```diff
@@ -107,6 +109,7 @@ public VanillaHandlerLookupService() {
 			bind(EncryptionKeyResponseMessage.class, EncryptionKeyResponseHandler.class);
 			bind(PlayerTabCompleteMessage.class, PlayerTabCompleteHandler.class);
 			bind(PlayerLocaleViewDistanceMessage.class, PlayerLocaleViewDistanceHandler.class);
+			bind(CommandBlockMessage.class, CommandBlockHandler.class);
```
Comment (zml2008) [inline src/main/java/org/spout/vanilla/plugin/protocol/VanillaHandlerLookupService.java]: Unnecessary

File: src/main/java/org/spout/vanilla/plugin/protocol/handler/world/block/CommandBlockHandler.java (hunk 3 above)
Comment (Snaipe) [inline src/main/java/org/spout/vanilla/plugin/protocol/handler/world/block/CommandBlockHandler.java]: It's fine, the input has no slash

File: src/main/java/org/spout/vanilla/plugin/protocol/codec/world/block/CommandBlockCodec.java (hunk 4 above)
Comment (Snaipe) [inline src/main/java/org/spout/vanilla/plugin/protocol/codec/world/block/CommandBlockCodec.java]: It has the same format as ChunkCacheCodec, it should be changed as well then

Comment (zml2008) [inline src/main/java/org/spout/vanilla/plugin/protocol/codec/world/block/CommandBlockCodec.java]: Yes it should.

File: src/main/java/org/spout/vanilla/plugin/component/substance/material/CommandBlockComponent.java [hunk 6]
```diff
@@ -0,0 +1,611 @@
...
+import org.spout.vanilla.api.event.block.CommandBlockUpdateEvent;
+
+public class CommandBlockComponent extends VanillaBlockComponent implements CommandSource {
+	
+	private final AtomicReference<ChatChannel> activeChannel = new AtomicReference<ChatChannel>();
+	
+	@Override
+	public void onAttached() {
+		super.onAttached();
+		this.activeChannel.set(Engine.STANDARD_BROADCAST_CHANNEL);
```
Comment (zml2008) [inline src/main/java/org/spout/vanilla/plugin/component/substance/material/CommandBlockComponent.java]: They should be in their own channel - permission-based channel with the permission vanilla.commandblock.broadcast

**Comment by windy1** [comment]
Can't be merged; needs a rebase.

**Comment by Snaipe** [comment]
@W1ndwaker done.

File: src/main/java/org/spout/vanilla/plugin/component/substance/material/CommandBlockComponent.java [hunk 7]
```diff
@@ -0,0 +1,612 @@
...
+	}
+
+	@Override
+	public Locale getPreferredLocale() {
+		return Locale.ENGLISH_US;
+	}
+	
+	public void executeCommandInput() {
+		String[] split = this.getCommandInput().split(" ");
+		if (split.length != 0)
```
Comment (windy1) [inline src/main/java/org/spout/vanilla/plugin/component/substance/material/CommandBlockComponent.java:135]: Formatting:

``` java
if (statement)
{
```

should be 

``` java
if (statement) {
```

File: src/main/java/org/spout/vanilla/plugin/component/substance/material/CommandBlockComponent.java [hunk 8]
```diff
@@ -0,0 +1,612 @@
...
+				
+				i += size + 1;
+				if (filter[i] != ',') {
+					return null;
+				}
+				
+				index++;
+			} else if (Character.isAlphabetic(filter[i])) {
+				String parameter;
+				{
```
Comment (windy1) [inline src/main/java/org/spout/vanilla/plugin/component/substance/material/CommandBlockComponent.java:208]:
``` java
int size = 0;
while (statement) {
    // do something
}
```

File: src/main/java/org/spout/vanilla/plugin/component/substance/material/CommandBlockComponent.java [hunk 9]
```diff
@@ -0,0 +1,612 @@
...
+				
+				if (i >= filter.length || filter[i] != '=') {
+					return null;
+				}
+				
+				int value;
+				{
+					i++;
+					int size = 0;
+					while (i + ++size < filter.length && Character.isDigit(filter[i + size]));
```
Comment (windy1) [inline src/main/java/org/spout/vanilla/plugin/component/substance/material/CommandBlockComponent.java:226]: See above

**Comment by windy1** [comment]
TBH, this seems a bit overcomplicated at first glance. I'll look into this further when I get home to see if we can simplify this a bit.

File: src/main/java/org/spout/vanilla/plugin/component/substance/material/CommandBlockComponent.java (hunk 8 above)
Comment (Snaipe) [inline src/main/java/org/spout/vanilla/plugin/component/substance/material/CommandBlockComponent.java:208]: I was using the bodiless while statement

**Comment by Snaipe** [comment]
I tried to have the fastest filter evaluation possible (the command block could be linked to a pulser), hence the fact I'm not using any regex and using more array manipulation - this might not be completely justified though, so feel free to change.

**Comment by Snaipe** [comment]
Also, I will probably be changing that part a little to allow plugins to register new fitlers when I'll have a little more time (ex: health or hunger filter), so I did the name/value parsing with the possibility of having a name longer than 2 chars.

**Comment by windy1** [comment]
Implemented in 33e160b2dcdb7d45243ef37184171ddafdf1d904

**Pull request closed without merging**